--- a/uqmi/uqmi.h
+++ b/uqmi/uqmi.h
@@ -129,2 +129,3 @@
 	uint8_t service;
+	uint16_t message;
 	uint16_t tid;
--- a/uqmi/dev.c
+++ b/uqmi/dev.c
@@ -125,2 +125,3 @@ static void qmi_process_msg(struct qmi_dev *qmi, struct qmi_msg *msg)
 	struct qmi_request *req;
+	struct qmi_request *next;
 	uint16_t tid;
@@ -378,3 +378,5 @@
-				while (!list_empty(&qmi->req)) {
-					req = list_first_entry(&qmi->req, struct qmi_request, list);
+				list_for_each_entry_safe(req, next, &qmi->req, list) {
+					if (req->service == QMI_SERVICE_CTL &&
+					    req->message == le16_to_cpu(sync_msg.ctl.message))
+						continue;
 					qmi_request_cancel(qmi, req);
@@ -478,2 +478,4 @@
 	req->service = msg->qmux.service;
+	req->message = req->service == QMI_SERVICE_CTL ?
+		le16_to_cpu(msg->ctl.message) : le16_to_cpu(msg->svc.message);
 	if (req->service == QMI_SERVICE_CTL) {
